finality-aleph/src/sync/select_chain.rs

```diff
@@ -1,111 +1,55 @@
-        info!(target: LOG_TARGET, "starting up StateWriter");
-        state_writer.run().await;
-        error!(target: LOG_TARGET, "StateWriter unexpectly finished");
-    });
-
-    handler
-}
-
-pub struct SelectChainStateHandler<Header> {
-    favourite_sender: Sender<Header>,
+#[derive(Clone)]
+pub struct FavouriteSelectChain<B: Block> {
+    favourite_block_request: mpsc::UnboundedSender<oneshot::Sender<B::Header>>,
 }
 
-impl<Header> SelectChainStateHandler<Header> {
-    pub fn new() -> (Receiver<Header>, Self) {
-        let (rx, tx) = channel(CHANNEL_SIZE as usize);
+impl<B: Block> FavouriteSelectChain<B> {
+    pub fn new() -> (Self, mpsc::UnboundedReceiver<oneshot::Sender<B::Header>>) {
+        let (rx, tx) = mpsc::unbounded();
 
         (
-            tx,
             Self {
-                favourite_sender: rx,
+                favourite_block_request: rx,
             },
+            tx,
         )
     }
-    pub fn update_favourite(&self, new_favourite: Header) {
-        if let Err(TrySendError::Full(_)) = self.favourite_sender.try_send(new_favourite) {
-            warn!(target: LOG_TARGET, "SelectChainStateHandler channel full, skipping one notification")
-        }
-    }
-}
-
-pub struct StateWriter<B: Block> {
-    events: Receiver<<B as Block>::Header>,
-    favourite: Arc<RwLock<<B as Block>::Header>>,
-}
-
-impl<B: Block> StateWriter<B> {
-    async fn run(&mut self) {
-        loop {
-            let new = self
-                .events
-                .recv()
-                .await
-                .expect("we hold the second end & its never ending stream.");
-            self.update_favourite(new);
-        }
-    }
-
-    fn update_favourite(&self, new_favourite: <B as Block>::Header) {
-        *self.favourite.write() = new_favourite;
-    }
-}
-
-#[derive(Clone)]
-pub struct FavouriteSelectChain<B: Block> {
-    favourite: Arc<RwLock<<B as Block>::Header>>,
-}
-
-impl<B: Block> FavouriteSelectChain<B> {
-    pub fn new(favourite: <B as Block>::Header) -> Self {
-        Self {
-            favourite: Arc::new(RwLock::new(favourite)),
-        }
-    }
 }
 
 #[async_trait::async_trait]
 impl<B, H> SelectChain<B> for FavouriteSelectChain<B>
 where
     B: BlockT<Header = H, Hash = BlockHash>,
     B: Block<Header = H, Hash = BlockHash>,
-    H: Sync + Send + Clone,
+    H: Sync + Send + Clone + Debug,
 {
     async fn leaves(&self) -> Result<Vec<<B as BlockT>::Hash>, Error> {
         // this is never used in the current version
-        Ok(vec![])
+        Ok(Vec::new())
     }
 
     async fn best_chain(&self) -> Result<<B as BlockT>::Header, Error> {
-        let best = self.favourite.read();
+        let (rx, tx) = oneshot::channel();
+
+        self.favourite_block_request
+            .unbounded_send(rx)
+            .expect("The second end of the channel should be operational");
+        let best = tx
+            .await
+            .expect("The second end of the channel should be operational");
```

Comment:
The two `expect`s here should just be errors.

Reply:
done